src/Executables/Game/Skills/PlayerSkills.cs

```diff
@@ -0,0 +1,597 @@
+            await _repository.SavePlayerSkillAsync(skill);
+        }
+    }
+
+    public ISKill? this[uint skillId] => _skills.TryGetValue(skillId, out var skill) ? skill : null;
+
+    public void SetSkillGroup(byte skillGroup)
+    {
+        if (skillGroup > 2) return;
+        if (_player.GetPoint(EPoints.Level) < 5) return;
+        
+        // todo: prevent changing skill group in certain situations
+        
+        _player.Player.SkillGroup = skillGroup;
+        
+        AssignDefaultActiveSkills();
+        
+        _player.Connection.Send(new QuantumCore.Game.Packets.Skills.ChangeSkillGroup
+        {
+            SkillGroup = skillGroup
+        });
+    }
+    
+    public void ClearSkills()
+    {
+        var points = _player.GetPoint(EPoints.Level) < 5
+            ? 0
+            : 4 + (_player.GetPoint(EPoints.Level) - 5) - _player.GetPoint(EPoints.Skill);
+        _player.SetPoint(EPoints.Skill, points);
+            
+        ResetSkills();
+    }
+    
+    public void ClearSubSkills()
+    {
+        var points = _player.GetPoint(EPoints.Level) < 10
+            ? 0
+            : (_player.GetPoint(EPoints.Level) - 9) - _player.GetPoint(EPoints.SubSkill);
+        
+        _player.SetPoint(EPoints.SubSkill, points);
+        
+        ResetSubSkills();
+    }
+
+    private void ResetSkills()
+    {
+        // store subskills in a temporary variable, clear the dictionary and then restore the subskills
+        var subSkills = _skills
+            .Where(sk => PassiveSkillIds.Contains(sk.Key))
+            .ToDictionary(sk => sk.Key, sk => sk.Value);
+        
+        _skills.Clear();
+        
+        if (_player.Player.SkillGroup > 0)
+        {
+            AssignDefaultActiveSkills();
+        }
+        
+        foreach (var subSkill in subSkills)
+        {
+            _skills[subSkill.Key] = subSkill.Value;
+        }
+
+        SendSkillLevelsPacket();
+    }
+
+    private void ResetSubSkills()
+    {
+        // Assign default values to passive skills
+        AssignDefaultPassiveSkills();
+
+        SendSkillLevelsPacket();
+    }
+
+    public void Reset(uint skillId)
+    {
+        if (skillId >= SkillMaxNum)
+        {
+            return;
+        }
+
+        if (!_skills.TryGetValue(skillId, out var skill)) return;
+
+        var level = skill.Level;
+        
+        skill.Level = 0;
+        skill.MasterType = ESkillMasterType.Normal;
+        skill.NextReadTime = 0;
+        
+        if (level > 17)
+            level = 17;
+        
+        _player.AddPoint(EPoints.Skill, level);
+
+        SendSkillLevelsPacket();
+    }
+
+    public void SetLevel(uint skillId, byte level)
+    {
+        if (skillId >= SkillMaxNum)
+        {
+            return;
+        }
+
+        if (!_skills.TryGetValue(skillId, out var skill)) return;
+
+        skill.Level = Math.Min((byte) 40, level);
+
+        skill.MasterType = level switch
+        {
+            >= 40 => ESkillMasterType.PerfectMaster,
```

Comment:
I think this should be configurable?

Reply:
I mean it can be configured for sure, but this is how the game client is expecting it to be. 
> i.e, a `MasterType` of `Normal` with a `Level` of `30` crashes my game client.

Reply:
Fair